File: sheet/prog/control_flow.js
"use strict";

let a = 10;
let fruits = ["apple", "banana", "fig"];

console.log("a"); console.log("b");

if (a < 10){
	console.log("less than 10.");
}else if(a == 10){
	console.log("equal to 10.");
}else{

}


// for
for (let i = 0; i < 5; i++){
	console.log(i);
}

// break and continue
OuterLoop: for (let i = 0; i < 5; i++){
	InnerLoop: for (let j = 0; j < 5; j++){
		if (j == 2){
			continue;
		}else if (j == 3){
			continue OuterLoop;
		}else if (j == 4){
			break;
		}
	}
}


// foreach
// for in:
//	オブジェクトの列挙可能(Enumerable)なプロパティを取得する。
//	予期せぬプロパティ(プロトタイプのものも含む)を取得してしまう可能性がある。
// for of [[2015]]:
//	Iterableなオブジェクト(配列を含む)から値を取得する。
for (let i in fruits){
	console.log(i)	// 0, 1, 2
	console.log(fruits[i])	// apple, banana, fig
}
for (let fruit of fruits){
	console.log(fruit)	// 
}
// map, forEachに関しても参照 TODO:

// while
while(a > 0){
	a--;
}
do{
	a++;
}while(a < 5);

// infinite loop
for (;;){
	break;
}
// switch
// 整数でなくても良い
switch(a){
	case 0:
	case 1:
		console.log("0 or 1");
		break;
	case 2:
		console.log("two");	// fallthrough
	default:
		console.log("other");
}

// goto
// ない

// node helloworld.js で実行

File: sheet/prog/control_flow.go
package main

import "fmt"
import "strconv"

func main() {
	a := 10
	fruits := []string{"apple", "banana", "fig"}

	// if
	if a < 10 {
		fmt.Println("less than 10.")
	} else if a == 10 {
		fmt.Println("equal to 10.")
	} else {

	}
	if i, _ := strconv.Atoi("2"); i == 2 {
		// 初期化節; 条件 も使用できる。
		fmt.Println("2")
	}

	// for
	for i := 0; i < 5; i++ {
		fmt.Println(i)
	}

	// for each (range clause)
	for i, f := range fruits {
		fmt.Println(i, f) // i -> index, f -> item
	}
	// range構文は適応するもので手に入るモノが違う
	// array/slice -> index, item
	// string -> index, rune
	// map -> key, value
	// channel -> element

	// break, continue
OuterLoop:
	for i := 0; i < 5; i++ {
	InnerLoop:
		for j := 0; j < 5; j++ {
			if j == 3 {
				continue
			} else if j == 4 {
				continue OuterLoop // ラベルを指定できる
			} else if j == 5 {
				break InnerLoop // breakも同様
			}
		}
	}

	// loop
	for a > 2 {
		a--
	}

	// infinite loop
	for {
		break
	}

	// Switch (初期化節も使用できる)
	// また、必ずしも整数でなくても良い (文字列でもOK)
	switch l := len("abc"); l {
		case 0, 1, 2, 3:
			// カンマで区切って値を指定できる。
			println("0-3")	// 標準でbreak不要！
		case 4:
			println("4")
			fallthrough		// fallthroutも指定可能
		default:
			println("other")
	}


	// interfaceのtype switch
	var iface interface {}
	iface = 3
	switch iface.(type) {
		case nil:
			println("nil")
		case int:
			println("int")
	}
	switch {
		case a == 0:
			println("zero")
		case a == 1:
			println("one")
			// if else チェーンの代わりにも使用できる。
	}

	


	goto MyLabel
MyLabel:
}
